From af2bf8b4c37dd08cfc75b50e82e4351803631eef Mon Sep 17 00:00:00 2001
From: Zhang Sanshan <b51434@freescale.com>
Date: Wed, 11 Mar 2015 21:16:55 +0800
Subject: [PATCH] MA-6382 enable wakelock when connect usb gadget

This patch prevent our board into suspend when connect usb device.

Signed-off-by: Zhang Sanshan <b51434@freescale.com>
---
 drivers/usb/gadget/android.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/gadget/android.c b/drivers/usb/gadget/android.c
index 97fdd1e4..daa0b17 100644
--- a/drivers/usb/gadget/android.c
+++ b/drivers/usb/gadget/android.c
@@ -27,6 +27,7 @@
 #include <linux/usb/ch9.h>
 #include <linux/usb/composite.h>
 #include <linux/usb/gadget.h>
+#include <linux/wakelock.h>
 
 #include "gadget_chips.h"
 
@@ -47,7 +48,7 @@ MODULE_LICENSE("GPL");
 MODULE_VERSION("1.0");
 
 static const char longname[] = "Gadget Android";
-
+static struct wake_lock wakelock;
 /* Default vendor and product IDs, overridden by userspace */
 #define VENDOR_ID		0x18D1
 #define PRODUCT_ID		0x0001
@@ -1517,6 +1518,7 @@ android_setup(struct usb_gadget *gadget, const struct usb_ctrlrequest *c)
 
 	spin_lock_irqsave(&cdev->lock, flags);
 	if (!dev->connected) {
+		wake_lock(&wakelock);
 		dev->connected = 1;
 		schedule_work(&dev->work);
 	} else if (c->bRequest == USB_REQ_SET_CONFIGURATION &&
@@ -1540,6 +1542,7 @@ static void android_disconnect(struct usb_composite_dev *cdev)
 
 	dev->connected = 0;
 	schedule_work(&dev->work);
+	wake_unlock(&wakelock);
 }
 
 static struct usb_composite_driver android_usb_driver = {
@@ -1605,6 +1608,7 @@ static int __init init(void)
 
 	_android_dev = dev;
 
+	wake_lock_init(&wakelock, WAKE_LOCK_SUSPEND, "gadget");
 	err = usb_composite_probe(&android_usb_driver);
 	if (err) {
 		pr_err("%s: failed to probe driver %d", __func__, err);
@@ -1629,6 +1633,7 @@ late_initcall(init);
 
 static void __exit cleanup(void)
 {
+	wake_lock_destroy(&wakelock);
 	usb_composite_unregister(&android_usb_driver);
 	class_destroy(android_class);
 	kfree(_android_dev);
-- 
1.8.0

